geomstats/stiefel.py

```diff
@@ -66,3 +70,284 @@ def random_uniform(self, n_samples=1):
+        if n_base_points == 1:
+            zeros = gs.zeros((n_tangent_vecs, p, p))
+        matrix_rz = gs.concatenate(
+            [matrix_r,
+             zeros],
+            axis=2)
+        block = gs.concatenate([matrix_ar, matrix_rz], axis=1)
+        matrix_mn_e = gs.expm(block)
+
+        if n_base_points == 1:
+            base_point = gs.tile(base_point, (n_tangent_vecs, 1, 1))
+        n_matrix_q = matrix_q.shape[0]
+        if n_matrix_q == 1:
+            matrix_q = gs.tile(matrix_q, (n_base_points, 1, 1))
+
+        exp = gs.einsum(
+            'nij,njk->nik',
+            gs.concatenate(
+                [base_point,
+                 matrix_q],
+                axis=2),
+            matrix_mn_e[:, :, 0:p])
+
+        return exp
+
+    def log(self, point, base_point, max_iter=100, tol=1e-6):
+        """
+        Riemannian logarithm of a point wrt a base point.
+
+        Based on:
+        Zimmermann, Ralf
+        "A Matrix-Algebraic Algorithm for the Riemannian Logarithm
+        on the Stiefel Manifold under the Canonical Metric"
+        SIAM J. Matrix Anal. & Appl., 38(2), 322–342.
+        """
+        point = gs.to_ndarray(point, to_ndim=3)
+        n_points, _, _ = point.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert (n_points == n_base_points
+                or n_points == 1
+                or n_base_points == 1)
+
+        if n_base_points == 1:
+            base_point = gs.tile(base_point, (n_points, 1, 1))
+        if n_points == 1:
+            point = gs.tile(point, (n_base_points, 1, 1))
+        n_logs = gs.maximum(n_base_points, n_points)
+
+        matrix_m = gs.matmul(gs.transpose(base_point, (0, 2, 1)), point)
+
+        # QR of normal component of a point
+        matrix_k = point - gs.matmul(base_point, matrix_m)
+
+        matrix_q = gs.zeros(matrix_k.shape)
+        matrix_n = gs.zeros((n_logs, p, p))
+        for i, k in enumerate(matrix_k):
+            matrix_q[i], matrix_n[i] = gs.linalg.qr(k)
+
+        # Orthogonal completion
+        matrix_w = gs.concatenate([matrix_m, matrix_n], axis=1)
+
+        matrix_v = gs.zeros((
+            matrix_w.shape[0],
+            max(matrix_w.shape[1], matrix_w.shape[2]),
+            max(matrix_w.shape[1], matrix_w.shape[2])
+            ))
+
+        for i, w in enumerate(matrix_w):
+            matrix_v[i], _ = gs.linalg.qr(w, mode="complete")
+
+        # Procrustes preprocessing
+        [matrix_d, matrix_s, matrix_r] = gs.linalg.svd(
+            matrix_v[:, p:2*p, p:2*p])
+
+        matrix_rd = gs.matmul(matrix_r, gs.transpose(matrix_d, axes=(0, 2, 1)))
+        matrix_v[:, :, p:2*p] = gs.matmul(matrix_v[:, :, p:2*p], matrix_rd)
+        matrix_v = gs.concatenate(
+            [gs.concatenate([matrix_m, matrix_n], axis=1),
+             matrix_v[:, :, p:2*p]],
+            axis=2)
+
+        for k in range(max_iter):
+
```

Comment:
line break - in general please make sure they are clean everywhere

Reply:
Done.